hal/network/lwip/cellular/pppncpnetif.cpp

```diff
@@ -99,41 +89,68 @@ if_t PppNcpNetif::interface() {
     return (if_t)client_.getIf();
 }
 
+void PppNcpNetif::setExpectedInternalState(NetifEvent ev) {
+    switch (ev) {
+        case NetifEvent::Up: {
+            expectedNcpState_ = NcpState::ON;
+            expectedConnectionState_ = NcpConnectionState::CONNECTED;
+            break;
+        }
+        case NetifEvent::Down: {
+            expectedConnectionState_ = NcpConnectionState::DISCONNECTED;
+            break;
+        }
+        case NetifEvent::PowerOff: {
+            expectedNcpState_ = NcpState::OFF;
+            expectedConnectionState_ = NcpConnectionState::DISCONNECTED;
+            break;
+        }
+        case NetifEvent::PowerOn: {
+            expectedNcpState_ = NcpState::ON;
+            break;
+        }
+    }
+}
+
 void PppNcpNetif::loop(void* arg) {
     PppNcpNetif* self = static_cast<PppNcpNetif*>(arg);
     unsigned int timeout = 100;
     while(!self->exit_) {
-        NetifEvent ev;
-        const int r = os_queue_take(self->queue_, &ev, timeout, nullptr);
         self->celMan_->ncpClient()->enable(); // Make sure the client is enabled
-        if (!r) {
-            // Event
-            LOG(TRACE, "PPP netif event from queue: %d", ev);
-            switch (ev) {
-                case NetifEvent::Up: {
-                    self->upImpl();
-                    break;
-                }
-                case NetifEvent::Down: {
-                    self->downImpl();
-                    break;
-                }
-                case NetifEvent::PowerOff: {
-                    self->downImpl();
-                    self->celMan_->ncpClient()->off();
-                    break;
-                }
-                case NetifEvent::PowerOn: {
-                    self->celMan_->ncpClient()->on();
-                    break;
-                }
+        os_semaphore_take(self->netifSemaphore_, timeout, false);
+
+        self->setExpectedInternalState(self->lastNetifEvent_);
```

Comment:
Without the queue now this may lose events. Imagine the following two calls:
```c++
Cellular.off();
Cellular.disconnect();
```

If for whatever reason the NCP netif thread doesn't have the time to act on both of these changes (e.g. it's blocked doing something in the meantime), we'll end up in just disconnected state with modem powered on.

Reply:
Good catch! 👍 The reason being is we don't set the expected state immediately on event requested.

Reply:
Yeah, my suggestion was to go with this approach only if there is a queue, but I didn't spell it out. If you do want to fix the original issue with sequential atomic operations an std::atomic struct would probably be better, the only caveat is that you'll need to modify it with compare_exchange_strong to make sure that the thread that modifies accounts for the changes other threads might have done.

Reply:
Or just leave things as they were, knowing that these events only come from system thread and it's fine to make sequential state changes.

Reply:
Alright, I'll leave things as-was and just add a comment in case that these APIs may be called by other thread.